R/estimateDominance.R

```diff
@@ -0,0 +1,366 @@
+        rank <- 1
+        as_relative <- TRUE
+    } else if (index %in% c("dmn")) {
+        #McNaughton's dominance
+        rank <- 2
+        as_relative <- TRUE
+        aggregate <- TRUE
+        #If index is "Simpson", calculates the Simpson to all the samples
+    } else if (index %in% c("simpson")) {
+        tmp <- apply(otu, 2, function(x) {
+            simpson_dominance(x)})
+        return(tmp)
+    } else if (index %in% c("core_abundance")) {
+        prevalence <- prevalentAbundance(x, detection=0, as_relative=TRUE)
+        return(prevalence)
+        #If index is "Gini" calculates the gini index to all the samples
+    } else if (index == "gini") {
+
+        gini <- gini(x)
+        return(gini)
+    }
+
+    if (rank == 1 && as_relative) {
+        index <- "dbp"
+    } else if (rank == 2 && as_relative && aggregate) {
+        index <- "dmn"
+    }
+
+    if (as_relative) {
+        #Calculates the relative abundance to all the columns
+        otu <- apply(otu, 2, function(x) {
+            x/sum(x, na.rm=TRUE)
+        })
+    }
+
+    #Aggregate or not
+    if (!aggregate) {
+        do <- apply(otu, 2, function(x) {
+            sort(x, decreasing = TRUE)[[rank]]
+        })
+    } else {
+        do <- apply(otu, 2, function(x) {
+            sum(sort(x, decreasing = TRUE)[seq_len(rank)])
+        })
+    }
+
+    #Add otus' names to the table
+    names(do) <- colnames(otu)
+
+    #Creates a matrix that has one column. Values of index is saved to that column.
+    if (is.vector(do)) {
+        do <- as.matrix(do, ncol=1)
+        colnames(do) <- index
+    }
+
+    #Returns a matrix that has one column. Rows are otus, column includes values of index.
+    do
+
+}
+
+# x: Species count vector
+simpson_dominance <- function(x, zeroes=TRUE) {
+
+    if (!zeroes) {
+        x[x > 0]
+    }
+
+    # Relative abundances
+    p <- x/sum(x)
+
+    # Simpson index (has interpretation as dominance)
+    lambda <- sum(p^2)
+
+    # More advanced Simpson dominance (Simpson 1949) However let us not use
+    # this as it is not in [0,1] and it is very highly correlated with the
+    # simpler variant lambda Species richness (number of species)
+    # S <- length(x) sum(p * (p - 1)) / (S * (S - 1))
+
+    lambda
+
+}
+
+gini <- function(x, abund_values="counts") {
+
+    #Stores the absolute abundances to "otu" variable
+    otu <- assays(x)[[abund_values]]
+
+    # Gini index for each sample
+    do <- apply(otu, 2, function(x) {
+        gini_help(x)
+    })
+    names(do) <- colnames(x)
+
+    do
+
+}
+
+
+
+gini_help <- function(x, w=rep(1, length(x))) {
+    # See also reldist::gini for an independent implementation
+    o <- order(x)
+    x <- x[o]
+    w <- w[o]/sum(w)
+    p <- cumsum(w)
+    nu <- cumsum(w * x)
+    n <- length(nu)
+    nu <- nu/nu[[n]]
+    sum(nu[-1] * p[-n]) - sum(nu[-n] * p[-1])
+}
+
+#Gets ME object and index data frame as parameters
+add_indices_to_coldata <- function(x, indicesDF, names){
```

Comment:
I think it is not working exactly same way. At least, something must change if that easier solution is used. 
In that point, indices have their own names i.e. "gini", "simpson"... In the ".add_indices_to_coldata" function, indices are stored in the object with the name that user has specified. It can be "gini", but it can also be "index1".

 colData(x) <- cbind(colData(x), indicesDF) does not take into account that. It just adds indices to the object. If user want that "gini" is "index1" it does not happen. 

I can change the functionality that way that user can not specify the name, so gini is always "gini". Or I can change it that way that the indices name, that user has specified, are stored as a name of indices earlier.

Reply:
Also, I tested it both ways, and I found out thay if user calculates same indices multiple times, same indices are stored in the object multiple times, if cbind() is used. 

In the add_indices_to_coldata, one index is saved only one time. Of course, if user wants it to be stored multiple times, user can give them different names. E.g. there can be "gini" and "gini2", but not "gini" and "gini".

Reply:
Then try something like this:

```
colData(x)[,names] <- indicesDF
```

If the names exist they will get overwritten, if not they will get attached as long `length(names) == ncol(indicesDF)`

Reply:
I implemented .add_dvrsty_values_to_colData from esimateDiversity, because I thought that there is no need to do it differently.